### PR TITLE
TST: Move fully passing tests from test_interval_new to test_interval

### PR DESCRIPTION
**Pull request by jschendel**
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Some of the `test_interval_new.py` tests are fully passing, so moving them to `test_interval.py` to ensure that new changes don't break them.

Added a couple of TODO's to the new tests related to checking non-monotonic/overlapping/dupe indexes, which generally weren't covered in the new or old tests (with one noted exception).  For the most part existing new/old tests just check the non-overlapping monotonic case.

Made some additional minor modifications to `test_interval_new.py`: cleaned up a couple tests by using `parametrize`, and fixed a few instances of a misplaced parenthesis.

### CODE REVIEW

File: pandas/tests/indexes/interval/test_interval.py [hunk 1]
```diff
@@ -400,16 +400,9 @@ def test_get_item(self, closed):
                                              closed=closed)
         tm.assert_index_equal(result, expected)
 
-    # To be removed, replaced by test_interval_new.py (see #16316, #16386)
+    # To be removed after overlapping/non-monotonic/duplicate cases have been
+    # added to test_get_loc_scalar below
```
Comment (jschendel) [inline pandas/tests/indexes/interval/test_interval.py]: The new `test_get_loc_scalar` only tests the monotonic non-overlapping case, so there are a few things in this test not fully covered.  Can add them to `test_get_loc_scalar` in this PR if desired, but didn't want to do so immediately since I didn't want to deviate too far from the existing tests.

File: pandas/tests/indexes/interval/test_interval.py [hunk 2]
```diff
@@ -482,11 +493,18 @@ def test_slice_locs_decreasing_int64(self):
...
-    def test_slice_locs_fails(self):
-        index = IntervalIndex.from_tuples([(1, 2), (0, 1), (2, 3)])
+    @pytest.mark.parametrize('query', [
+        [0, 1], [0, 2], [0, 3], [3, 1], [3, 4], [0, 4]], ids=str)
+    @pytest.mark.parametrize('tuples', [
+        [(0, 2), (1, 3), (2, 4)],  # increasing overlapping
+        [(2, 4), (1, 3), (0, 2)],  # decreasing overlapping
+        [(0, 2), (0, 2), (2, 4)],  # sorted duplicates
+        [(0, 2), (2, 4), (0, 2)],  # unsorted duplicates
+        [(0, 2), (0, 2), (2, 4), (1, 3)]])  # another unsorted duplicates
```
Comment (jschendel) [inline pandas/tests/indexes/interval/test_interval.py]: Slight modification of the deleted test to parametrize over tuples.  Makes the test code itself cleaner.

File: pandas/tests/indexes/interval/test_interval_new.py [hunk 3]
```diff
@@ -21,45 +21,25 @@ def _compare_tuple_of_numpy_array(self, result, expected):
 
     @pytest.mark.parametrize("idx_side", ['right', 'left', 'both', 'neither'])
     @pytest.mark.parametrize("side", ['right', 'left', 'both', 'neither'])
-    def test_get_loc_interval(self, idx_side, side):
-
+    @pytest.mark.parametrize("bound", [
+        [0, 1], [1, 2], [2, 3], [3, 4], [0, 2], [2.5, 3], [-1, 4]], ids=str)
```
Comment (jschendel) [inline pandas/tests/indexes/interval/test_interval_new.py]: the entire test was within a `for` loop, pulled it out of the test by using `parametrize`

File: pandas/tests/indexing/interval/test_interval.py [hunk 4]
```diff
@@ -51,9 +60,8 @@ def test_getitem_with_scalar(self):
         tm.assert_series_equal(expected, s[s >= 2])
 
     # TODO: check this behavior is consistent with test_interval_new.py
-    @pytest.mark.parametrize('direction, closed',
-                             product(('increasing', 'decreasing'),
-                                     ('left', 'right', 'neither', 'both')))
+    @pytest.mark.parametrize('direction', ['increasing', 'decreasing'])
+    @pytest.mark.parametrize('closed', ['left', 'right', 'neither', 'both'])
```
Comment (jschendel) [inline pandas/tests/indexing/interval/test_interval.py]: minor unrelated cleanup; seems simpler just have two instances of `parametrize`

File: pandas/tests/indexing/interval/test_interval_new.py [hunk 5]
```diff
@@ -214,18 +186,6 @@ def test_loc_with_overlap(self):
         with pytest.raises(KeyError):
             s.loc[1:4]
 
-    def test_non_unique(self):
```
Comment (jschendel) [inline pandas/tests/indexing/interval/test_interval_new.py]: this test looks identical to the existing one, so just deleted it entirely

File: pandas/tests/indexes/interval/test_interval.py [hunk 6]
```diff
@@ -420,8 +413,26 @@ def test_get_loc_value(self):
         assert idx.get_loc(3) == 1
         pytest.raises(KeyError, idx.get_loc, 3.5)
 
-        idx = IntervalIndex.from_arrays([0, 2], [1, 3])
-        pytest.raises(KeyError, idx.get_loc, 1.5)
+    # TODO: add overlapping/non-monotonic/duplicate index tests
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: can you add these with an xfail (or if there are a lot, leave then in _new)

File: pandas/tests/indexes/interval/test_interval.py [hunk 7]
```diff
@@ -420,8 +413,26 @@ def test_get_loc_value(self):
...
+    @pytest.mark.parametrize("scalar", [-0.5, 0, 0.5, 1, 1.5, 2, 2.5, 3, 3.5])
+    def test_get_loc_scalar(self, closed, scalar):
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: can you put errors in a separate test (test_get_loc_scalar_error)

File: pandas/tests/indexes/interval/test_interval_new.py [hunk 8]
```diff
@@ -21,45 +21,25 @@ def _compare_tuple_of_numpy_array(self, result, expected):
...
-                else:
-                    with pytest.raises(KeyError):
-                        idx.get_loc(Interval(*bound, closed=side))
+        # if get_loc is supplied an interval, it should only search
+        # for exact matches, not overlaps or covers, else KeyError.
+        if idx_side == side:
+            if bound == [0, 1]:
+                assert idx.get_loc(Interval(0, 1, closed=side)) == 0
+            elif bound == [2, 3]:
+                assert idx.get_loc(Interval(2, 3, closed=side)) == 1
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval_new.py]: can the error tests be pulled out to a separate test

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/interval/test_interval.py (hunk 7 above)
Comment (jschendel) [reply pandas/tests/indexes/interval/test_interval.py]: Yes, but imo doing so would be less concise add code bloat.  Could certainly do it, but want to make the downsides known first.

By separating them, you can't really parametrize over both `scalar` and `closed`, since testing endpoints will raise for some values of `closed` but pass for others, so one of them will need to be written inside a for loop of some kind.  It'd also add a bit of repetitiveness, since you'd need to encode passing keys in the passing test, and failing keys in the errors tests, whereas the current version only encodes passing with the missing values presumed to be failing.

File: pandas/tests/indexes/interval/test_interval_new.py (hunk 8 above)
Comment (jschendel) [reply pandas/tests/indexes/interval/test_interval_new.py]: Yes, with somewhat similar tradeoffs described above, but I don't think it'd be as bad.  This seems like it'd be more reasonable to split, but haven't fully scoped it out.

**Comment by jschendel** [comment]
Refactored the `get_loc` tests for scalars and intervals, and added additional cases to test (overlapping, non-monotonic with dupes).  The scalar case is still fully passing.

Haven't split the errors out into separate tests yet; wanted to see if the refactoring makes things better first.  Seems like splitting the errors out would add some bloat, but can do it if that would be preferred.

File: pandas/tests/indexes/interval/test_interval.py [hunk 9]
```diff
@@ -400,28 +400,72 @@ def test_get_item(self, closed):
...
+
+        Parameters
+        ----------
+
+        idx: IntervalIndex
+            IntervalIndex over which get_loc checks will be run
+
+        key: scalar or Interval
+
+        correct: dict
```
Comment (jreback) [inline pandas/tests/indexes/interval/test_interval.py]: not a fan of this. quickly these test routines get very bespoke, and are not well known and hence repeated often enough. This is doing too much; should not catch raises. Its *ok* that it can compare an array and scalar though.

Maybe if you make a general one thats a function (and not restricted to comparisions here) would be ok.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jschendel** [comment]
@jreback : what are your thoughts on the testing approach in the new file I just pushed, `test_interval_specific.py`? Haven't made any changes other than adding this file; didn't want to do too much in case I'm heading down the wrong path.

The idea is to use this approach to test behavior that depends on the underlying structure of the intervals in the index (e.g. overlapping vs. nonoverlapping monotonic), such as `get_loc`, which has specific code paths for different scenarios like this.  Behavior that is independent from the underlying interval structure (e.g. `IntervalIndex.mid`) can be left in the existing `test_interval.py` setting.

Was a little complicated to get all of this set up, but shouldn't be too bad now if someone else wanted to create a new test class variation, e.g. `TestMonotonicDecreasing`.  Not at all tied to the name `test_interval_specific.py` either; could even see this being split off into separate files for each child class, e.g. `test_interval_overlapping.py`.

File: pandas/tests/indexes/interval/test_interval_specific.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+class TestNonOverlappingMonotonic(Base):
+    """Tests specifics to non-overlapping monotonic IntervalIndex"""
+    # tuples from which to create the underlying IntervalIndex
+    tuples = [(0, 1), (2, 3)]
+
+    # valid scalar tuples to test, of the from (closed abbr, scalar)
+    # (a scalar is valid if it's contained in an interval in the index)
+    scalar_tuples_valid = [
+        ('l', 0), ('l', 0.5), ('l', 2), ('l', 2.5),
```
Comment (jschendel) [inline pandas/tests/indexes/interval/test_interval_specific.py]: Could remove some of these to eliminate redundancy, e.g. `('l', 0.5)` and `('l', 2.5)` are testing fundamentally the same thing.

**Comment by jreback** [comment]
it’s well documented and pretty clear once u get pass the initial magic

@jorisvandenbossche @TomAugspurger

**Comment by jschendel** [comment]
Closing this PR for the time being.  Not all that happy with this setup, as it seems a bit over-engineered.  Might try something similar later one once some of the `IntervalIndex` methods and API have been solidified.
